### PR TITLE
separate DatetimeIndex timezone tests

### PR DESCRIPTION
**Pull request by jbrockmendel**
There is probably some overlap with other tests.indexes.datetimes modules.  We'll get those in the next pass.

### CODE REVIEW

File: pandas/tests/indexes/datetimes/test_timezones.py [hunk 1]
```diff
@@ -0,0 +1,1047 @@
...
+        return self.__offset
+
+    def tzname(self, dt):
+        return self.__name
+
+    def dst(self, dt):
+        return timedelta(0)
+
+
+fixed_off = FixedOffset(-420, '-07:00')
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_timezones.py]: make these into fixtures

File: pandas/tests/indexes/datetimes/test_timezones.py [hunk 2]
```diff
@@ -0,0 +1,1047 @@
...
+        for freq, n in [('H', 1), ('T', 60), ('S', 3600)]:
+            idx = date_range(datetime(2011, 3, 26, 23),
+                             datetime(2011, 3, 27, 1), freq=freq)
+            idx = idx.tz_localize('UTC')
+            idx = idx.tz_convert('Europe/Moscow')
+
+            expected = np.repeat(np.array([3, 4, 5]), np.array([n, n, 1]))
+            tm.assert_index_equal(idx.hour, Index(expected))
+
+    def test_dti_tz_convert_dst(self):
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_timezones.py]: parameterize (can be next pass)

File: pandas/tests/indexes/datetimes/test_timezones.py [hunk 3]
```diff
@@ -0,0 +1,1047 @@
...
+        dti = DatetimeIndex(start='11/6/2011 1:59', end='11/6/2011 2:00',
+                            freq='L')
+        with pytest.raises(pytz.AmbiguousTimeError):
+            dti.tz_localize(tzstr)
+
+        dti = DatetimeIndex(start='3/13/2011 1:59', end='3/13/2011 2:00',
+                            freq='L')
+        with pytest.raises(pytz.NonExistentTimeError):
+            dti.tz_localize(tzstr)
+
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_timezones.py]: prob should dedupe these tz parameters into some nicely named fixtures (and push up what you can to conftest)

Comment (jreback) [reply pandas/tests/indexes/datetimes/test_timezones.py]: next PR

File: pandas/tests/indexes/datetimes/test_timezones.py [hunk 4]
```diff
@@ -0,0 +1,1047 @@
...
+        # When there is no dst transition, nothing special happens
+        dr = date_range(datetime(2011, 6, 1, 0), periods=10,
+                        freq=pd.offsets.Hour())
+        is_dst = np.array([1] * 10)
+        localized = dr.tz_localize(tz)
+        localized_is_dst = dr.tz_localize(tz, ambiguous=is_dst)
+        tm.assert_index_equal(localized, localized_is_dst)
+
+    # TODO: belongs outside tz_localize tests?
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_timezones.py]: yeah move to construction

File: pandas/tests/indexes/datetimes/test_timezones.py [hunk 5]
```diff
@@ -0,0 +1,1047 @@
...
+            expected = date_range('1/1/2000', periods=10, freq='D',
+                                  tz=tzlocal())
+            tm.assert_index_equal(result, expected)
+
+            assert result.is_normalized
+            assert not rng.is_normalized
+
+    # ------------------------------------------------------------
+    # DatetimeIndex.__new__
+
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_timezones.py]: hmm maybe should be in construction

File: pandas/tests/indexes/datetimes/test_timezones.py [hunk 6]
```diff
@@ -0,0 +1,1047 @@
...
+        idx2 = DatetimeIndex(start="2005-11-10 08:00:00", freq='H', periods=2,
+                             tz=tzstr)
+        idx3 = DatetimeIndex(arr, tz=tzstr)
+        idx4 = DatetimeIndex(np.array(arr), tz=tzstr)
+
+        for other in [idx2, idx3, idx4]:
+            tm.assert_index_equal(idx1, other)
+
+    # ------------------------------------------------------------
+    # Arithmetic; may belong in test_arithmetic
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_timezones.py]: yes

**Review by jreback — CHANGES_REQUESTED** [review]
needs a rebase

File: pandas/tests/indexes/datetimes/test_timezones.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/datetimes/test_timezones.py]: They're not being passed into parametrize anywhere.

File: pandas/tests/indexes/datetimes/test_timezones.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/datetimes/test_timezones.py]: Yah I parametrized the most egregious of these, but there is bunch of cleanup like this for the next pass.

File: pandas/tests/indexes/datetimes/test_timezones.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/datetimes/test_timezones.py]: See the "FIXME" note about 4 lines down.  I'm not confident this test is testing what it is intending. This is where the existing indentation typo is: https://github.com/pandas-dev/pandas/pull/19545/files#diff-f67205b3c64b63fa7471c4221e413a19L424

File: pandas/tests/indexes/datetimes/test_timezones.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/indexes/datetimes/test_timezones.py]: hmm, these should be then.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks. if you can add the above items to the master TODO list would be great. I know minor, but don't want to forget about them; at the same time no reason to hold up merging this.

**Comment by jbrockmendel** [comment]
Will do.  I appreciate the leeway.

**Pull request merged**